text/0018-generalized-in.md

```diff
@@ -0,0 +1,185 @@
+# Generalized `in` operator
+
+## Related issues and PRs
+
+- Reference Issues:
+- Implementation PR(s):
+
+## Timeline
+
+- Start Date: 2023-07-11
+- Date Entered FCP:
+- Date Accepted:
+- Date Landed:
+
+## Summary
+
+Allow `in` to be used for set membership, not just hierarchy membership. In particular, allow it to be used with non-entity-typed values on the LHS. The detailed design of this proposal (see below) ensures that back-compatibility is maintained for all existing valid Cedar policies.
+
+## Basic examples
+
+### Example 1
+Today's Cedar:
+```
+permit(principal, action == Action::"view", resource == Document::"mine.doc")
+when {
+    ["sales", "legal"].contains(principal.department)
+};
+```
+
+Proposed:
+```
+permit(principal, action == Action::"view", resource == Document::"mine.doc")
+when {
+    principal.department in ["sales", "legal"]
+};
+```
+
+### Example 2
+Today's Cedar:
+```
+permit(
+    principal == Principal::"long-uuid-here",
+    action == Action::"view",
+    resource in Database::"prod"
+) when {
+    resource has tags
+        && resource.tags has tagA
+        && ["value1", "value2"].contains(resource.tags.tagA)
+};
+```
+
+Proposed:
+```
+permit(
+    principal == Principal::"long-uuid-here",
+    action == Action::"view",
+    resource in Database::"prod"
+) when {
+    resource has tags
+        && resource.tags has tagA
+        && resource.tags.tagA in ["value1", "value2"]
+};
+```
+
+## Motivation
+
+In today's Cedar, `.contains()` is the only way to check set membership. There are two problems with this:
+1. This is simply awkward in some use cases, such as the two examples above. Subjectively, the `.contains()` syntax is hard to read in these cases, because the set is required to be on the LHS and the value you're actually interested in is required to be on the RHS. In these examples, having the set on the RHS instead of the LHS reads more intuitively.
+2. Anecdotally, many users intuitively expect that `in` can be used for set membership and not just hierarchy membership, and are confused when this results in Cedar type errors. See, for example, [this thread](https://cedar-policy.slack.com/archives/C0547KH7R19/p1689253188448809) on the Cedar Slack.
```

Comment:
Do we have any examples of customers who want the semantics we are actually proposing? The slack thread you mention is confusion about which operator to use: The user wanted exactly what `.contains()` is providing, semantically, but mistakenly used operator `in` instead. A better error message could direct them to the right construct.

I'm dubious of making a serious language change to make it "easier" for people to write code. This sort of choice was made with many scripting languages (I'm looking at you, Javascript), and the results are not always positive. 

Do we have examples where a customer wants `principal.foo in ["a", "b", User::"foo"]` to be `==` for the first two, and deep `in` for the last? If yes, that's more support for this proposal. If no, what would customers think should happen? If they mostly guess right, that's more support. If they don't, then you are possibly introducing a subtle bomb in the semantics.

Reply:
The reason for this proposal instead of a simpler semantics like Alternative 5, is that it can be done backwards-compatibly, while Alternative 5 is a breaking change.  If we think the breaking change is acceptable for Cedar 3, we can consider Alternative 5 and, as @khieta also says on a different thread here, the semantics would be simpler and more intuitive.  However we haven't so far greenlit any breaking changes to the Cedar policy syntax itself (at least to my memory), and I'm hesitant to do that, even in a "new major version" like Cedar 3.  This particular breaking change would be particularly insidious because it wouldn't be that previously-valid policies now produce parse errors, evaluation errors, or even validation errors, but rather that the behavior simply silently changes out from under you when you upgrade to Cedar 3.

Reply:
I think Alternative 5 is too invasive at this point, even with a major version change.  This is a fundamental change to the semantics of a core operator.

I think we should make the error message better / more informative, and then wait to see if a change to semantics is needed, and if so, which alternative meets most users’ expectations.

Reply:
re: breaking changes in Cedar policy syntax for Cedar 3 -- I've now (somewhat hypocritically) proposed one in [RFC 20](https://github.com/cedar-policy/rfcs/pull/20), but I believe there are significant mitigating factors in that case that are not present with Alternative 5.  In RFC 20's case, the break is that previously valid Cedar policies would now produce errors, while in Alternative 5, previously valid Cedar policies would now silently have a different non-error behavior, which in my opinion is worse.  Also, I suspect the number of users who have Cedar policies in the class that RFC 20 would break, is significantly smaller than the number of users who have Cedar policies in the class that Alternative 5 would break.

Reply:
I've updated the discussion of Alternative 5 to note the concerns about backward compatibility mentioned in this thread.  Since this thread (5 months ago), we have also significantly improved the error message, and I've adjusted the RFC text to note this as well.  As a result, I will resolve this thread.